Update transport default preference to use min number of transports

ron-murhammer opened the pull request:
Addresses point 5 on #553 - Update transport default preference to use min number of transports with min remaining capacity

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/triplea-game/triplea/739)
<!-- Reviewable:end -->

DanVanAtta commented on src/games/strategy/triplea/util/TransportUtils.java:
```diff
@@ -67,22 +67,37 @@
-          it.remove();
+
+      // Check if remaining units can all be loaded into 1 transport
+      final int capacity = TransportTracker.getAvailableCapacity(transport);
```
Ideally cleanup changes are in their own commit, makes for easier reviewing. (marked variable as final)

ron-murhammer replied:
Yeah, I originally had several commits for this but the merge didn't go smoothly so I ended up redoing the changes in 1 new commit.

DanVanAtta replied:
understood, but did want to point out the review difficulty.

DanVanAtta commented on src/games/strategy/triplea/util/TransportUtils.java:
```diff
@@ -198,4 +213,18 @@ public int compare(final Unit o1, final Unit o2) {
     return Optional.empty();
   }
 
+  private static void loadMaxUnits(final Unit transport, final List<Unit> canBeTransported,
```
This is a bit procedural. The data state of `mapping` is being controlled by code that does not own the `mapping` object. So that is the problem that one would want to fix. In this case we would have pure in parameters and an out parameter. This way the data is still owned by the client of this method, we just compute a value for them here.

Another notable thing is that this code is repeated:
`final int capacity = TransportTracker.getAvailableCapacity(transport);`
That makes me think it has been sliced up not so cleanly.

I think the way to go is to get the first unit with cost less then capacity. Then have the client code deal with updating the map.

Something like this:
`canBeTransported.stream().filter( unit -> UnitAttachment.get((unit).getType()).getTransportCost() <= capacity ).getFirst()`

That returns an optional, which would then play well with the optional above.

ron-murhammer replied:
Agree that its not great but I don't think there is a good way to really make it an input/output method while still trying to pull the low level details out of the calling method. I think keeping the calling method simple and clean is more important in this case.

DanVanAtta replied:
The calling method is a bit better, but we've also taken a step back as well. This method is simply and effectively not in-lined anymore. That does not help so much since you need to go down and still read this method, it does not succeed in any information hiding. The small step back, we have some code duplication in the method.

The reason for writing tests is to force a clean design, I recommend taking that approach, so there is some real pressure to have a clean, testable design. 

Failing any updates, if I were to start refactoring this, an eventual step I would probably take would be to inline this method, eliminate the duplicated data.

DanVanAtta replied:
Ultimately a pivot of the code is probably needed, from procedural to OO. IE: perhaps a transport object that could help absorb some of the complexity, in combination with a unit object, with methods like: `unit.canFitOnTransport(transport)`, and should work with a way to sort the list of units by their transport cost.

ron-murhammer replied:
This method is called in 2 places so inlining it will cause the calling method to become very large and difficult to deal with. `getAvailableCapacity` is a very light method that needs called at least twice even if you in line it since 2 different transport objects are passed. Without a major change this code isn't easy to make much cleaner. Since I have PRs based on this same class and this code is still better than 99% of the code in the project I'd like to move forward. If you want to spend some time looking at the best way to refactor and give more detail then we can do it in a follow on PR.

DanVanAtta commented on src/games/strategy/triplea/util/TransportUtils.java:
```diff
@@ -133,7 +148,7 @@ public static int getTransportCost(final Collection<Unit> units) {
     final Collection<Unit> canBeTransported = Match.getMatches(units, Matches.UnitCanBeTransported);
     final Collection<Unit> canTransport = Match.getMatches(transports, Matches.UnitCanTransport);
-    final Map<Unit, Unit> mapping = new HashMap<>();
+    final Map<Unit, Unit> mapping = new HashMap<Unit, Unit>();
```
Bad merge resolution?

ron-murhammer replied:
Yep.

DanVanAtta commented on src/games/strategy/triplea/util/TransportUtils.java:
```diff
@@ -67,22 +67,37 @@
       }
+
+      // Load max units into current transport
+      loadMaxUnits(transport, canBeTransported, mapping);
```
consider renaming `transport` to `currentTransport`, and remove the comment that is just restating the code. 
"// Load max units into current transport"

ron-murhammer replied:
Done.

DanVanAtta replied:
Github is good now about removing comments that are out of data, and showing new commits. I prefer 'done' comments to be avoided, let the commit do that job for you. As is, I have no evidence this was actually done.

DanVanAtta commented on src/games/strategy/triplea/util/TransportUtils.java:
```diff
@@ -67,22 +67,37 @@
+      final int remainingCost = getTransportCost(canBeTransported);
+      if (remainingCost <= capacity) {
+        if (!finalTransport.isPresent() || capacity < TransportTracker.getAvailableCapacity(finalTransport.get())) {
+          finalTransport = Optional.of(transport);
```
` loadMaxUnits` here and break, then drop the Optional. It's the last transport you need, it does not seem like there is any point in continuing wit the loop anymore.

ron-murhammer replied:
That won't work. Notice the if is OR not AND. You want to find the transport with minimum remaining capacity that can fit all remaining units. Since they are sorted by max -> min size you need to loop through them all.

DanVanAtta replied:
If they are sorted like that, why not loop from the end to beginning? In that case you do not need to loop through them all.

ron-murhammer replied:
Loop needs to be sorted with the largest transport first since we are trying to use min number of transports. The exception to this is the last transport being loaded. Example is say we have a small transport (holds 1) and a big transport (holds 2) and we have 1 unit left, we want to load the unit in the small transport as this wastes the minimum amount of space.

DanVanAtta commented on src/games/strategy/triplea/util/TransportUtils.java:
```diff
@@ -67,22 +67,37 @@
-          mapping.put(unit, transport);
-          it.remove();
+
+      // Check if remaining units can all be loaded into 1 transport
```
These two comments effectively only have the for loop in between, 2 lines apart essentially, and they don't really seem to agree.
```
// Add max units to each transport
for (final Unit transport : canTransport) {
// Check if remaining units can all be loaded into 1 transport
```

ron-murhammer replied:
How so? The first is describing what the loop is doing at a high level (load max units into each transport). The second is describing what the first chunk of code in the loop is doing (checking if remaining units can be loaded all into 1 transport).

DanVanAtta replied:
One states "each" transport, then we are doing something that could only impact "1 transport", with an obvious loop break that would prevent us from doing something to "each" transport.

Both comments indicate serious smells, as they paraphrase what the code is doing, which should be self-evident.

Basically it's as if the second comment is restating the definition of the for loop... which is really not helpful. A "first check if we can load all units into 1 transport" would be better, but that still reads weird with a seemingly different game plan stated two lines earlier.

DanVanAtta replied:
Add max units to each transport, means if we have 3 units, and 3 transports, we load 1 unit in each transport.

But "remaining units can all be loaded into 1 transport", means that with 3 units and 3 transports, we load the remaining 3 into just 1 if they will fit

So the comments are not consistent with how the code functions.

ron-murhammer replied:
"Add max units to each transport" means add max units to each transport so if you have 3 units then try to add *max* (3 in this case) to the largest transport. This means use the least number of transports possible not the most number of transports.

ron-murhammer commented:
@DanVanAtta I think I addressed all your comments and made a few changes. I'd like to get this merged without major changes since the next transport PR is based on this and touches much of the same code.

DanVanAtta commented on src/games/strategy/triplea/util/TransportUtils.java:
```diff
@@ -67,22 +67,37 @@
+        if (!finalTransport.isPresent() || capacity < TransportTracker.getAvailableCapacity(finalTransport.get())) {
+          finalTransport = Optional.of(transport);
         }
+        continue; // skip transports until the smallest one all units fit in with the most moves
```
The comment on the continue does not make sense

ron-murhammer replied:
Updated this comment for clarity.

DanVanAtta commented on src/games/strategy/triplea/util/TransportUtils.java:
```diff
@@ -72,14 +72,14 @@
-          finalTransport = Optional.of(transport);
+          finalTransport = Optional.of(currentTransport);
         }
         continue; // skip transports until the smallest one all units fit in with the most moves
```
This is not addressed, comment still does not make sense.

ron-murhammer replied:
Reworded comment.

DanVanAtta commented:
Still a couple of comments to address @ron-murhammer

ron-murhammer commented:
@DanVanAtta All comments should be addressed now and updated one of the code comments for clarity.

DanVanAtta commented:
For the loop with the optional final transport, you stated:
"Since they are sorted by max -> min size you need to loop through them all."

If we have that sorting, then could we iterate from the end of the list to the front to find the right transport sooner? Currently we have a foreach to loop from the front to end.

DanVanAtta commented:
> "Add max units to each transport" means add max units to each transport so if you have 3 units then try to add max (3 in this case) to the largest transport. This means use the least number of transports possible not the most number of transports.

I read "Add max units to each transport"  as meaning to add the most units possible across every  transport, effectively to spread them out. That was the original functionality of this loop too, no?
As it stands the phrase is ambiguous, and is put in confusing contrast with the new commentary added. I recommend deleting the line.

ron-murhammer commented:
@DanVanAtta Comment removed. 

You would need to resort the list in order to go in reverse order since its also sorted by movement and you want the one with least capacity and most remaining moves.

No that is the method above it. The method name is `mapTransportsToLoadUsingMinTransports` which I think clearly indicates you want to put max units into min number of transports.
